core/server/proxy/src/test/java/alluxio/StreamCacheTest.java

```diff
@@ -34,11 +38,184 @@ public void operations() throws Exception {
+        mCacheID = cacheID;
+        mNullable = nullable;
+      }
+
+      @Override
+      public void run() {
+        FileInStream cachedInStream = streamCache.getInStream(mCacheID);
+        if (mNullable && cachedInStream == null) {
+          return;
+        }
+        Assert.assertSame(is, cachedInStream);
+      }
+    }
+
+    class GetOutputStream implements Runnable {
+      private int mCacheID;
+      private boolean mNullable;
+
+      GetOutputStream(int cacheID, boolean nullable) {
+        mCacheID = cacheID;
+        mNullable = nullable;
+      }
+
+      @Override
+      public void run() {
+        FileOutStream cachedOutStream = streamCache.getOutStream(mCacheID);
+        if (mNullable && cachedOutStream == null) {
+          return;
+        }
+        Assert.assertSame(os, cachedOutStream);
+      }
+    }
+
+    class InvalidateInputStream implements Runnable {
+      private int mCacheID;
+
+      InvalidateInputStream(int cacheID) {
+        mCacheID = cacheID;
+      }
+
+      @Override
+      public void run() {
+        Assert.assertSame(is, streamCache.invalidate(mCacheID));
+      }
+    }
+
+    class InvalidateOutputStream implements Runnable {
+      private int mCacheID;
+
+      InvalidateOutputStream(int cacheID) {
+        mCacheID = cacheID;
+      }
+
+      @Override
+      public void run() {
+        Assert.assertSame(os, streamCache.invalidate(mCacheID));
+      }
+    }
+
+    final int cacheSize = 100;
+    final List<Thread> threads = new ArrayList<>();
+
+    // Concurrent put.
+    // Sequentially run PutInputStream and PutOutputStream.
+    final AtomicIntegerArray inputStreamCacheIDs = new AtomicIntegerArray(cacheSize);
+    final AtomicIntegerArray outputStreamCacheIDs = new AtomicIntegerArray(cacheSize);
+    for (int i = 0; i < cacheSize; i++) {
+      threads.add(new Thread(new PutInputStream(inputStreamCacheIDs, i)));
+      threads.add(new Thread(new PutOutputStream(outputStreamCacheIDs, i)));
+    }
+    for (Thread thread : threads) {
+      thread.start();
+    }
+    for (Thread thread : threads) {
+      thread.join();
+    }
+    Assert.assertEquals(cacheSize * 2, streamCache.size());
+    for (int i = 0; i < cacheSize; i++) {
+      Assert.assertSame(is, streamCache.getInStream(inputStreamCacheIDs.get(i)));
+      Assert.assertSame(os, streamCache.getOutStream(outputStreamCacheIDs.get(i)));
+    }
+
+    // Concurrent get.
+    // Sequentially run GetInputStream and GetOutputStream.
+    threads.clear();
+    for (int i = 0; i < cacheSize; i++) {
+      threads.add(new Thread(new GetInputStream(inputStreamCacheIDs.get(i), false)));
+      threads.add(new Thread(new GetInputStream(outputStreamCacheIDs.get(i), false)));
+    }
+    for (Thread thread : threads) {
+      thread.start();
+    }
+    for (Thread thread : threads) {
+      thread.join();
+    }
+    Assert.assertEquals(cacheSize * 2, streamCache.size());
+
+    // Concurrent put, get, and invalidate.
```

Comment:
I think that the following organization would accomplish similar coverage and does not introduce subtle dependencies between different parts of the test:

```
final StreamCache streamCache = new StreamCache(Constants.HOUR_MS);
final FileInStream is = Mockito.mock(FileInStream.class);
final FileOutStream os = Mockito.mock(FileOutStream.class);

class CacheTest<T> implements Runnable {
  private AtomicIntegerArray mCacheIDs;
  private int mIndex;
  private T mStream;

  StreamCacheRunnable(T stream, AtomicIntegerArray cacheIDs, int index) {
    mCacheIDs = cacheIDs;
    mIndex = index;
    mStream = stream;
  }

  @Override
  public void run() {
    mCacheIDs.set(mIndex, streamCache.put(mStream));
    Assert.assertSame(mStream, streamCache.getInStream(mCacheIDs.get(mIndex)));
    Assert.assertSame(mStream, streamCache.invalidate(mCacheIDs.get(mIndex)));
  }
}

final int numOps = 100;
final List<Thread> threads = new ArrayList<>();

final AtomicIntegerArray inputStreamCacheIDs = new AtomicIntegerArray(numOps);
final AtomicIntegerArray outputStreamCacheIDs = new AtomicIntegerArray(numOps);
for (int i = 0; i < numOps; i++) {
  threads.add(new Thread(new CacheTest<InputStream>(inputStreamCacheIDs, i, is)));
  threads.add(new Thread(new CacheTest<OutputStream>(outputStreamCacheIDs, i, os)));
}
for (Thread thread : threads) {
  thread.start();
}
for (Thread thread : threads) {
  thread.join();
}
```